pkg/yurtappmanager/controller/nodepool/nodepool_controller.go

```diff
@@ -0,0 +1,600 @@
+	oldNp := oldNode.Labels[appsv1alpha1.LabelCurrentNodePool]
+
+	if newNp != oldNp {
+		if newNp == "" {
+			// remove node from old pool
+			klog.V(5).Infof("will enqueue old pool(%s) for node(%s)",
+				oldNp, newNode.GetName())
+			addNodePoolToWorkQueue(oldNp, q)
+			return
+		}
+
+		if oldNp == "" {
+			// add node to the new Pool
+			klog.V(5).Infof("will enqueue new pool(%s) for node(%s)",
+				newNp, newNode.GetName())
+			addNodePoolToWorkQueue(newNp, q)
+			return
+		}
+		klog.V(5).Infof("will enqueue both new pool(%s) and"+
+			" old pool(%s) for node(%s)",
+			newNp, oldNp, newNode.GetName())
+		addNodePoolToWorkQueue(oldNp, q)
+		addNodePoolToWorkQueue(newNp, q)
+		return
+	}
+
+	if isNodeReady(*newNode) != isNodeReady(*oldNode) {
+		// if the newNode and oldNode status are different
+		klog.V(5).Infof("node phase has been changed,"+
+			" will enqueue pool(%s) for node(%s)", newNp, newNode.GetName())
+		addNodePoolToWorkQueue(newNp, q)
+		return
+	}
+
+	if !reflect.DeepEqual(newNode.Labels, oldNode.Labels) ||
+		!reflect.DeepEqual(newNode.Annotations, oldNode.Annotations) ||
+		!reflect.DeepEqual(newNode.Spec.Taints, oldNode.Spec.Taints) {
+		// if node's labels, annotations or taints are updated
+		// TODO only consider the pool realted attributes
+		klog.V(5).Infof("nodepool related attributes has been changed,"+
+			" will enqueue pool(%s) for node(%s)",
+			newNp, newNode.GetName())
+		addNodePoolToWorkQueue(newNp, q)
+	}
+
+}
+
+// Delete implements EventHandler
+func (e *EnqueueNodePoolForNode) Delete(evt event.DeleteEvent,
+	q workqueue.RateLimitingInterface) {
+	node, ok := evt.Object.(*corev1.Node)
+	if !ok {
+		klog.Error("fail to assert runtime Object to v1.Node")
+		return
+	}
+
+	np := node.Labels[appsv1alpha1.LabelCurrentNodePool]
+	if np == "" {
+		klog.V(5).Infof("node(%s) doesn't belong to any pool", node.GetName())
+		return
+	}
+	// enqueue the nodepool that the node belongs to
+	klog.V(5).Infof("will enqueue pool(%s) as node(%s) has been deleted",
+		np, node.GetName())
+	addNodePoolToWorkQueue(np, q)
+}
+
+// Generic implements EventHandler
+func (e *EnqueueNodePoolForNode) Generic(evt event.GenericEvent,
+	q workqueue.RateLimitingInterface) {
+	return
+}
+
+// createNodePool creates an nodepool, it will retry 5 times if it fails
+func createNodePool(c client.Client, name string,
+	poolType appsv1alpha1.NodePoolType) {
+	for i := 0; i < 5; i++ {
+		np := appsv1alpha1.NodePool{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: appsv1alpha1.NodePoolSpec{
+				Type: poolType,
+			},
+		}
+		err := c.Create(context.TODO(), &np)
+		if err == nil {
+			klog.V(4).Infof("the default nodepool(%s) is created", name)
+			break
+		}
+		if apierrors.IsAlreadyExists(err) {
+			klog.V(4).Infof("the default nodepool(%s) already exist", name)
+			break
+		}
+		klog.Errorf("fail to create the node pool(%s): %s", name, err)
+		time.Sleep(2 * time.Second)
+	}
```

Comment:
You should add a log if the default node pool fails to be created in the end or `exit` the controller since something goes wrong.

Reply:
output a V4 log message if we fail to create the default nodepool after trying 5 times.